src/Microsoft.Security.Utilities.Core/CompiledHighPerformancePattern.cs

```diff
@@ -0,0 +1,285 @@
+    /// signficant byte. However, this encoding should be considered an
+    /// implementation detail that can change.
+    /// </remarks>
+    public int PackedSignature { get; }
+
+    /// <summary>
+    /// Gets the pre-computed pattern for the given signature or null if there is none.
+    /// </summary>
+    public static CompiledHighPerformancePattern? ForSignature(string signature)
+    {
+        return s_patternsBySignature.TryGetValue(signature, out CompiledHighPerformancePattern? pattern) ? pattern : null;
+    }
+
+    /// <summary>
+    /// Regex options used for all high-performance regexes.
+    /// </summary>
+    /// <remarks>
+    /// We don't use 'NonBacktracking' here because it is not supported by the
+    /// regex source generator. However these high-performance patterns should
+    /// not cause problematic backtracking. They are extremely simple, anchored
+    /// to the front with at most one optional group at the end. Generated code
+    /// has been inspected to see that the only "loops" have at most 1
+    /// iteration. If more complex regexes are added, the generated code should
+    /// be re-examined.
+    ///
+    /// 'SingleLine' is used to ensure that '.' matches any character including
+    /// new lines. This is used to share regexes that can purely skip over the
+    /// signature that has already been matched before the regex is applied.
+    ///
+    /// Also note that the high-performance scanner does not allow substitution
+    /// of the regex engine and the regex engine built-in to .NET is always
+    /// used. Profiling showed no performance gain from RE2, even on .NET
+    /// Framework, even for the unoptimized 'SecretMasker' using only the
+    /// highly-identifiable patterns, but this merits more scrutiny and
+    /// investigation.
+    ///
+    /// Even more critically, the regex engine abstraction does not currently
+    /// provide a mechanism to search within a substring without allocation.
+    /// There is no incentive to add it, which would require work in the RE2
+    /// wrapper, unless further investigation reveals a flaw in the finding on
+    /// RE2 performance above.
+    /// </remarks>
+    private const RegexOptions Options = RegexOptions.Compiled |
+                                         RegexOptions.ExplicitCapture |
+                                         RegexOptions.CultureInvariant |
+                                         RegexOptions.Singleline;
+
+
+    public CompiledHighPerformancePattern(string signature,
+                                           int signaturePrefixLength,
+                                           int minMatchLength,
+                                           int maxMatchLength,
+                                           Regex scopedRegex)
+    {
+#if HIGH_PERFORMANCE_CODEGEN // This validation happens at compile time and needn't be repeated at runtime.
+        if (!scopedRegex.ToString().StartsWith("^"))
```

Comment:
>scopedRegex [](http://example.com/codeflow?start=13&length=11)

why is this important? nice to comment on in code later. also, the 3 or 4 char limit, which would 5 chars be unhelpful for example. note that other projects declare lengthier signatures for general IndexOf check, will we need limited fix sig expression to generalize this class for those scenarios?

Reply:
Both of these can be relaxed, and I'm sure they will be when I look at #156.

I didn't want the cost of doing things that rust identifiable_scan does not do to show up in the benchmark here.

Reply:
More details.

If the regexes are not `^` scoped then they cannot have a fixed signature-prefix-length. I suspect I will change this to *max*-signature-prefix-length and only the rules that actually have fixed length will use `^`.

If rules have variable pre-signature length, then we also can't play the trick of using the same regex that skips over the signature.

If I try to explain all this in a comment, I will surely make a bunch more typos, so I'll do it in a follow-up.